docs(pki): Add section for pki web config (#12861)
Scille/parsec-cloud: docs/hosting/deployment/idopte-pki-web.rst

.. Parsec Cloud (https://parsec.cloud) Copyright (c) BUSL-1.1 2016-present Scille SAS

.. cspell:words pkiweb

.. _doc_hosting_deployment_idopte_pkiweb:

.. |idopte| replace:: `Idopte`_

.. _Idopte: https://www.idopte.com/

=====================================

> Enable PKI Web with Idopte middleware

Reviewer: This section should be renamed and moved as a sub-section at the end of the existing webapp-deploy.rst file. 

```suggestion
Enable PKI and SmartCard support
```

=====================================

> Configure the :ref:`web application <doc_hosting_deployment_webapp>` to enable PKI integration using |idopte|'s middleware.

Reviewer: ```suggestion
This section describes how to enable PKI and SmartCard support for the :ref:`web application <doc_hosting_deployment_webapp>` using |idopte|'s *Smart Card Middleware*.
```

> |idopte| provide a middleware installed on the user's device that provide the Smart Card Web Services (SCWS).

Reviewer: I would use "computer" instead of "device" (so there is no confusion with Parsec devices). 

```suggestion
|idopte| *Smart Card Middleware*, installed on the user's computer, provides the *Smart Card Web Services* (SCWS)
that are necessary to communicate with a smart card from the web browser.
```

Why using Idopte service?

> =========================

Reviewer: ```suggestion
Why Idopte?
===========
```

We use Idopte to provide the integration of smartcard features on the web application because at the moment of writing the `Web Smart Card API <https://wicg.github.io/web-smart-card/>`_ is in draft.

> And the client that asked for the support of smartcard in web was already using that solution with other web applications.

Reviewer 2: I'm not sure this remark is useful here. And maybe it's a bit on the paranoid side but if someone has our list of clients and idopte's one they could draw conclusions.

> And the client that asked for the support of smartcard in web was already using that solution with other web applications.

Reviewer: I would add some info from Idopte's documentation:

```suggestion
There is no standard way to access the smart card hardware from the web browser.
A couple of alternatives were considered, among which we can mention the 
`Web Smart Card API <https://wicg.github.io/web-smart-card/>`_ but, at the time of writing, the API is in draft.

We chose Idopte, a French software company specialized in Smart Card Middleware,
mainly because:

- It provides a simple and functional JavaScript API allowing communication
  with the smart card from the browser (through their middleware)
- The Parsec customer needing web smart card support is already using Idopte
  middleware with other web applications.
```

.. note::
   :collapsible: open

>    We're open to additional integration.

Reviewer: ```suggestion
   We are open to consider other options for smart card web support.
```

>    But the proposed integration has to provide the following features:

Reviewer: ```suggestion
   Please feel free to contact us if you know of any other alternatives 
   supporting following features:
```

   - List available certificate (at least their DER contents)

>    - Provide a way to verify the trust in a certificate (either providing the CRL list or doing the check itself)

Reviewer: ```suggestion
   - Check certificate trust (either providing :abbr:`CRL (Certificate revocation list)` or doing the check itself)
```

   - Allow to sign/verify data
   - Allow to encrypt/decrypt data

Prerequisite

> ============

Reviewer: ```suggestion
Prerequisites
=============
```

> Before starting to configure the PKI Web feature, ensure that you fill the following prerequisite:

Reviewer: ```suggestion
Before you begin, make sure you meet the following prerequisites:
```

.. _scws_mutual_auth: https://www.idopte.com/doc_scws.php

- You have an RSA Private key.

  .. tip::
     :collapsible: open

     You can generate a private key with ``openssl``:

     .. code-block:: bash

        openssl genrsa -out idopte-service.key 4096

> - |idopte| provided you with a web certificate that include the public key of the private key above.

Reviewer: ```suggestion
- |idopte| provided you a web certificate including the public key of the RSA private key above.
```

  .. tip::
     :collapsible: open

     ``openssl`` can help you generate the public key:

     .. code-block:: bash

        openssl rsa -in idopte-service.key -out idopte-service.pub

> - On top of the web certificate, you have obtained from Idopte their public keys that are used during the `mutual authentication process <scws_mutual_auth_>`_

Reviewer: ```suggestion
- You have Idopte's public keys, used during the `mutual authentication process <scws_mutual_auth_>`_
```

> - You have a copy of the patched SCWS javascript file

Reviewer: ```suggestion
- You have a copy of the patched ``scwsapi.js`` file
```

  .. important::

>      You cannot use a non-patched SCWS javascript file else it will fail to be loaded by the application.

Reviewer: ```suggestion
     You cannot use a non-patched ``scwsapi.js`` file else the web application will not be able to load it.
```

  .. note::

>      We cannot distribute that file patched because of license reason

Reviewer: ```suggestion
     For legal reasons, we are currently unable to distribute the patched ``scwsapi.js`` file.
```

Configure the webapp
====================

> The web application need some configuration to have the feature enabled:

Reviewer: ```suggestion
Follow these steps to configure the web application:
```

> #. You need to copy the SCWS javascript file into the assets folder:

Reviewer: ```suggestion
#. Copy the ``scwsapi.js`` file into the assets folder:
```

   .. tip::

>       For better caching purpose, we recommend to rename the SCWS script file to have it's checksum in the name:

Reviewer: ```suggestion
      To improve caching, we recommend renaming the ``scwsapi.js`` file to have it's checksum in the name:
```

      .. code-block:: bash

         mv scwsapi.js scwsapi-$(sha256sum scwsapi.js | cut -c -6).js

> #. And edit some `meta tags <https://developer.mozilla.org/en-US/docs/Web/HTML/Reference/Elements/meta>`_ in the ``index.html`` file:

Reviewer: ```suggestion
#. Edit the following `meta tags <https://developer.mozilla.org/en-US/docs/Web/HTML/Reference/Elements/meta>`_ in the ``index.html`` file:
```

   - Put the web certificate provided by |idopte| in the content of the meta tag named ``scws-web_application_certificate``
   - Edit the content of the meta tag with the name ``scws-scwsapi_js-location`` to path from where the SCWS javascript file is served

Configure the server

> ====================

Reviewer: ```suggestion
Configure Parsec Server
=======================
```

> The server need additional configuration to be able to `perform the mutual authentication with the client middleware <scws_mutual_auth_>`_:

Reviewer: ```suggestion
The server needs additional configuration to be able to `perform the mutual authentication with the client middleware <scws_mutual_auth_>`_:
```

- Make |idopte|'s public keys accessible to the server.

>   Put the keys into a file and either set the option ``--scws-idopte-public-keys-file`` or the env variable ``PARSEC_SCWS_IDOPTE_PUBLIC_KEYS_FILE`` to the location where that file reside.

Reviewer: ```suggestion
  Put the keys into a file and either set the option ``--scws-idopte-public-keys-file`` or the env variable ``PARSEC_SCWS_IDOPTE_PUBLIC_KEYS_FILE`` 
  to the file location.
```

- Provide the private key to use:

  .. important::

>      The private key is expected to be provide in it's PEM format

Reviewer 2: ```suggestion
     The private key is expected to be provide in its PEM format
```

>   For that you can either provide the key in a file via the option ``--scws-web-application-private-key-file`` or the env variable ``PARSEC_SCWS_WEB_APPLICATION_PRIVATE_KEY_FILE``.

Reviewer: ```suggestion
  For that you can either specify the key file via the ``--scws-web-application-private-key-file`` option 
  or the env variable ``PARSEC_SCWS_WEB_APPLICATION_PRIVATE_KEY_FILE``.
```

  Or via it's content in base64 with the env variable ``PARSEC_SCWS_WEB_APPLICATION_PRIVATE_KEY_CONTENT``

  .. code-block:: bash

     export PARSEC_SCWS_WEB_APPLICATION_PRIVATE_KEY_CONTENT=$(base64 idopte-service.key)

  .. warning::

>      The server also provide the option ``--scws-web-application-private-key-content`` but it's not recommended to use it as the private key value will then be present in the shell history.

Reviewer: ```suggestion
     The server also provides the ``--scws-web-application-private-key-content`` option but it's not recommended 
     because the private key will then be present in the shell history.
```
